### PR TITLE
CBAPI-4251: Add Dynamic Criteria support to Asset Groups

### PR DESCRIPTION
**Pull request by abowersox-cb**
## Pull request checklist

Please check if your PR fulfills the following requirements:

- [x] Docs have been reviewed and added / updated if needed (for bug fixes / features)
- [x] Tests have been added that prove the fix is effective or that the feature works.
- [x] New and existing tests pass locally with the changes.
- [x] Code follows the style guidelines of this project (PEP8, clean code, linter).
- [x] A self-review of the code has been done.

## What is the ticket or issue number?

https://jira.carbonblack.local/browse/CBAPI-4251

## Pull Request Description

Added dynamic criteria support (the `query` attribute) to the asset groups.

## Does this introduce a breaking change?

- [ ] Yes
- [x] No

## How Has This Been Tested?

Unit tests have been modified to use the `query` attribute throughout, especially when testing `create_group()`. Coverage for the file stands at 97%.

### CODE REVIEW

File: src/cbc_sdk/platform/asset_groups.py [hunk 1]
```diff
@@ -75,34 +95,41 @@ def _query_implementation(cls, cb, **kwargs):
...
         Required Permissions:
-            gm.group-set (CREATE)
+            gm.group-set(CREATE)
 
         Args:
             cb (BaseAPI): Reference to API object used to communicate with the server.
             name (str): Name for the new asset group.
             description (str): Description for the new asset group.
             policy_id (int): ID of the policy to be associated with this asset group.
+            query (str): Query string to be used to dynamically populate this group. Default is ``None``.
```
Comment (avanbrunt-cb) [inline src/cbc_sdk/platform/asset_groups.py:110]: Can we indicate that `None` requires manual assignment to add devices to the group

**Review by avanbrunt-cb — APPROVED** [review]

File: src/cbc_sdk/platform/asset_groups.py (hunk 1 above)
Comment (abowersox-cb) [reply src/cbc_sdk/platform/asset_groups.py:110]: Yes.

File: src/cbc_sdk/platform/asset_groups.py [hunk 2]
```diff
@@ -75,34 +95,41 @@ def _query_implementation(cls, cb, **kwargs):
...
 
         Returns:
             AssetGroup: The new asset group.
         """
-        group_data = {"name": name, "description": description, "member_type": "DEVICE", "policy_id": policy_id}
+        group_data = {"name": name, "description": description, "member_type": "DEVICE",
```
Comment (kebringer-cb) [inline src/cbc_sdk/platform/asset_groups.py:116]: Policy_id and query are both optional.
If query is null, it needs to be excluded from the request.
This is a valid create group request body:
{
    "description": "Testing new SDK",
    "member_type": "DEVICE",
    "name": "SDK Test 01"
}

File: src/cbc_sdk/platform/asset_groups.py [hunk 3]
```diff
@@ -11,7 +11,22 @@
...
+An *asset group* represents a group of devices (endpoints, VM workloads, and/or VDIs) that can have a single policy
+applied to it so the protections of all similar assets are synchronized with one another.  Policies carry a "position"
+value as one of their attributes, so that, between the policy attached directly to the device, and the policies
+attached to any asset groups the device is a member of, the one with the highest "position" is the one that applies to
+that device.  Devices may be added to an asset group either explicitly, or implicitly by specifying a query on the
+asset group, such that all devices matching that search criteria are considered part of the asset group.
+
+Typical usage example:
+
+    # assume "cb" is an instance of CBCloudAPI
```
[the remark this answers is not in the document]
Comment (kebringer-cb) [reply src/cbc_sdk/platform/asset_groups.py:26]: RTD link on this PR:
https://carbon-black-cloud-python-sdk--446.org.readthedocs.build/en/446/cbc_sdk.platform/#module-cbc_sdk.platform.asset_groups

[the remark this answers is not in the document]
Comment (abowersox-cb) [reply src/cbc_sdk/platform/asset_groups.py:26]: "Typical usage example:" needs a _double_ colon at the end. Fixing this.

File: src/cbc_sdk/platform/asset_groups.py [hunk 4]
```diff
@@ -20,18 +35,23 @@
...
     swagger_meta_file = "platform/models/asset_group.yaml"
 
     def __init__(self, cb, model_unique_id=None, initial_data=None, force_init=False, full_doc=False):
         """
-        Initialize the AssetGroup object.
+        Initialize the ``AssetGroup`` object.
 
         Required Permissions:
-            gm.group-set (READ)
+            gm.group-set(READ)
```
[the remark this answers is not in the document]
Comment (abowersox-cb) [reply src/cbc_sdk/platform/asset_groups.py:54]: Updating all instances of `gm.group-set` to `group-management`.

**Review by kebringer-cb — APPROVED** [review]

**Pull request merged**